### PR TITLE
Add chunking and size related metrics in the router

### PR DESCRIPTION
**Pull request by pnarayanan**
- Histograms for blob sizes and number of chunks for puts and gets.
- Counter for blob types (simple/composite) for puts and gets.

### CODE REVIEW

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java [hunk 1]
```diff
@@ -104,6 +104,17 @@
   public Gauge<Long> chunkFillerThreadRunning;
   public Gauge<Long> requestResponseHandlerThreadRunning;
   public Gauge<Integer> activeOperations;
+  public Gauge<Integer> maxPutChunkSizeGauge;
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: this need not be declared as a class variable

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java [hunk 2]
```diff
@@ -277,6 +298,21 @@ public Integer getValue() {
...
+   * @param maxPutChunkSize the value to initialize this metric with.
+   */
+  public void initializeMaxPutChunkSizeMetric(final int maxPutChunkSize) {
+    maxPutChunkSizeGauge = new Gauge<Integer>() {
+      @Override
+      public Integer getValue() {
+        return maxPutChunkSize;
+      }
+    };
+    metricRegistry.register(MetricRegistry.name(NonBlockingRouter.class, "MaxPutChunkSizeBytes"), maxPutChunkSizeGauge);
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: I don't understand the purpose of this metric - the configuration should be enough to indicate this (in that sense it isn't even a metric - what are we measuring?). Why have a gauge that exports the same number over and over again - something that has _absolutely_ no chance of changing?

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java [hunk 3]
```diff
@@ -217,6 +228,16 @@ public NonBlockingRouterMetrics(ClusterMap clusterMap) {
     crossColoSuccessCount =
         metricRegistry.counter(MetricRegistry.name(NonBlockingRouter.class, "CrossColoSuccessCount"));
 
+    // metrics to track blob sizes and chunking.
+    putBlobSizeBytes = metricRegistry.histogram(MetricRegistry.name(PutManager.class, "PutBlobSizeBytes"));
+    putBlobChunkCount = metricRegistry.histogram(MetricRegistry.name(PutManager.class, "PutBlobChunkCount"));
+    getBlobSizeBytes = metricRegistry.histogram(MetricRegistry.name(PutManager.class, "GetBlobSizeBytes"));
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: wrong owner class

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java [hunk 4]
```diff
@@ -217,6 +228,16 @@ public NonBlockingRouterMetrics(ClusterMap clusterMap) {
...
+    getBlobChunkCount = metricRegistry.histogram(MetricRegistry.name(PutManager.class, "GetBlobChunkCount"));
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: wrong owner class

**Comment by vgkholla** [comment]
Done with a pass. Looks mostly good.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java (hunk 2 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: This value can change between invocations of the router, if we play around with chunk sizes. As mentioned in the description: though this would not change for a running router, this metric would help maintain historical information if the chunk sizes are changed. We could probably get rid of this eventually, but I think it will be useful at least in the interim. This should be harmless as the only cost is that of the gauge itself - no computation costs. Is there a better way to track this information?

**Comment by pnarayanan** [comment]
Addressed comments.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java [hunk 5]
```diff
@@ -105,6 +105,16 @@
   public Gauge<Long> requestResponseHandlerThreadRunning;
   public Gauge<Integer> activeOperations;
 
+  // metrics for tracking blob sizes and chunking.
+  public final Histogram putBlobSizeBytes;
```
Comment (nsivabalan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: guess we have followed the pattern of putBlobSizeInBytes in our code base. ("In")

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java [hunk 6]
```diff
@@ -277,6 +297,20 @@ public Integer getValue() {
   }
 
   /**
+   * Initializes a {@link Gauge} metric for the maxPutChunkSize. This value is a constant for a running router
+   * instance, but can change between invocations. This metric helps maintain historical information.
+   * @param maxPutChunkSize the value to initialize this metric with.
+   */
+  public void initializeMaxPutChunkSizeMetric(final int maxPutChunkSize) {
```
Comment (nsivabalan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: Why is this required. We will have history to our config files right and I don't think this might change frequently either.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java [hunk 7]
```diff
@@ -105,6 +105,16 @@
   public Gauge<Long> requestResponseHandlerThreadRunning;
   public Gauge<Integer> activeOperations;
 
+  // metrics for tracking blob sizes and chunking.
+  public final Histogram putBlobSizeBytes;
+  public final Histogram putBlobChunkCount;
+  public final Histogram getBlobSizeBytes;
+  public final Histogram getBlobChunkCount;
+  public final Counter simpleBlobPutCount;
```
Comment (nsivabalan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: Is the intention was counter or rate ? Can we start using Meters to track rates instead of counters?

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java (hunk 2 above)
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: If we play around with chunk sizes, won't the commit history have the details we need? 
It is not harmful because of the computation cost (which is minimal) but because of the idea that a configuration has to be exported as a metric to keep track of history of a certain config - you can present the same point for _any_ configuration value.

Also keep in mind that the metric itself is ephemeral - sure you can have another product that maintains the history of the metric (just like commit histories) but the metric itself is not useful for the purpose you mention.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java (hunk 5 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: This is consistent with rest of the metrics in the router (see the time related metrics).

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java (hunk 7 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: Intention is to track the count.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java (hunk 2 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: makes sense. removed.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java (hunk 6 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: removed.

**Comment by pnarayanan** [comment]
Addressed comments. When you merge, please ensure that the statement about the gauge is removed from the description.

**Comment by vgkholla** [comment]
LGTM. @nsivabalan Please take a look at your addressed comments. I will merge after.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java (hunk 7 above)
Comment (nsivabalan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: So, this will be ever increasing count?

Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: Just want to note here that even `Meter` has an ever increasing count but also exports a rate - so in essence it provides the info of a `Counter` + some extra data. Personally I think `Counter` is right here.

**Comment by nsivabalan** [comment]
LGTM

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java (hunk 7 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: yes, `Meter` exports additional information including rates and moving averages at different intervals. All that is an overkill for this metric. This is no different from the error count metrics.
